compose: Disable "Message X" button for DMs when DMs are not allowed.

N-Shar-ma opened the pull request:
When in a private narrow, the "Message X" button is disabled if direct messages are not allowed in the organisation and the current recipient is not a bot.

Note that when the recipient is a user group with 1, more or all bots, the button is disabled then too as such PMs are not allowed. Only when the recipient is a single bot, then it's not disabled, as DMs with one bot are allowed even in organisations where DMs are disabled.

**Screenshots and screen captures:**
![image](https://user-images.githubusercontent.com/68962290/226442307-575511fd-6b5e-4ade-be50-f955ec0cfbf5.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Nice! Could we also disable the `x` keyboard shortcut?

alya commented:
Oh, and this also wasn't in the issue description, but I think the "Message X" button should also be disabled in PM conversations (group or 1:1) with non-bot users.

![Screen Shot 2023-01-02 at 10 22 26 AM](https://user-images.githubusercontent.com/2090066/210267470-309a8387-8309-48d0-bd52-7affb253ad4c.png)

N-Shar-ma commented:
@alya Have implemented both those features! Here's how the disabled "Message X" button looks:
![image](https://user-images.githubusercontent.com/68962290/210316424-c6c6cedf-3c93-423f-82a0-f321a589097d.png)
![image](https://user-images.githubusercontent.com/68962290/210316477-b645eb5a-5a61-4396-b72b-eb7dc72a9136.png)

The commit message additionally explains why it's not disabled for user groups with bots

alya commented:
Looks good to me in manual testing!

@amanagr could you please review this one?

alya commented:
I also filed #23970 to address the resulting difficulty of PMing a bot when PMs are disabled; we may want to do something in the personal settings bot panel as well.

amanagr commented:
LGTM!

alya commented:
@timabbott This PR has been approved by @amanagr and me -- over to you.

palashb01 commented:
Hello, @alya! Since my PR, [#23899](https://github.com/zulip/zulip/pull/23899), shares a lot of common code with this one, I'm waiting for this PR to merge before I update my PR, Thanks. (FYI: @timabbott)

timabbott commented:
Generally we want any permissions rule check like this to get in a function in `settings_data.js` -- `settings_data.user_can_send_private_messages()`. This is really good for helping make sure the checks are accurate, and becomes pretty important as we make changes to how permissions work.

(The check for that situation will become more complicated as we refine our settings related to private message policies.)

That said, given [our plans](https://chat.zulip.org/#narrow/stream/101-design/topic/restricting.20PMs/near/1504451) in this space I think this feature may not end up being helpful -- it's going to be exceedingly rare that you actually can't PM anyone (and you can always PM yourself or bots, and this would be a way to do that). So I think we should skip the "New private message" button part of this.

The last commit, which has a specific recipient, I think does make sense -- and for that, I think what we want is instead to add a function `settings_data.user_can_direct_message(user_ids_array))` that handles all the cases -- for now just the setting and whether the recipients are all bots, and have that function be used for the "Message ..." button's disabled state. (We'll then end up replacing that function as we adjust the design).

Or maybe we want a slightly different API that lets the shared function return the error string (or `null` if it's approved), I'm not sure -- but the logic for deciding if the request is allowed should be centralized in one place in the `settings_data` module.

N-Shar-ma commented:
@timabbott Got it, but the `settings_data` module is in typescript and we can't import and use some functionalities necessary for the `user_can_direct_message` fuction from the javascript module `people` to check if recipients are all bots. 🤔

timabbott commented:
Sorry for the slow reply -- I think we can just put the function for whether we can send a direct message to a given recipient in `people.js`.

N-Shar-ma commented:
@timabbott I have refactored the last commit, and removed the other two as discussed, and also added tests; this should be ready for another review from you now!

alya commented:
@timabbott Re-adding the integration review label, as your feedback above has been addressed.

timabbott commented:
I think we should do tippy tooltips rather than an HTML title element (complete with the nice keyboard shortcut hints); we can compute the tooltip content on hover, which may simplify the `update_buttons` code path significantly.

But that seems like a fine follow-up, so merged after rewriting the core `user_can_direct_message` function, thanks @N-Shar-ma!

N-Shar-ma commented:
@timabbott The auto-merge did not work since a puppeteer test was failing as there was an error trying to set the state of the Compose message button when narrowed to a non-existing user. I have now fixed that, so please re-review before merging :)

timabbott commented:
OK, merged, thanks @N-Shar-ma! If you have time to Tippy tooltip migration suggested in https://github.com/zulip/zulip/pull/23944#issuecomment-1492498966 in time for 7.0, that'd be awesome; can you open an issue for the task so we can tag it as a release goal (I don't want us to forget about it)?

N-Shar-ma commented:
@timabbott opened issue #25096, though zulipbot didn't let me tag it as release-goal

alya commented:
That's because the label doesn't have a dash in it. :)
